File: src/shared/sources/source-ledger.ts
/**
 * What the sources screen needs to say about a source — pure, DOM-free.
 *
 * Three jobs, none of which belongs in a component:
 *
 * - **A health state and the sentence that goes with it.** A connection is
 *   not a boolean: an instance can answer, refuse (401 — sign in *there*),
 *   or never answer at all, and each of those asks for a different verb. The
 *   mapping from a `WinnowError` to that state is the same everywhere, so it
 *   lives here rather than in the screen that happens to show it first.
 * - **How many documents a source holds.** Forgetting a connection strands
 *   every trip and project whose `sourceId` is that host, so the screen must
 *   be able to count them BEFORE it asks — "8 projects and 3 trips live
 *   there" is the difference between a decision and a surprise.
 * - **A short label for a host.** `winnow.steeve.website` in a 288px library
 *   tab truncated to `WINNOW.STEEVE.…`, which reads as a defect. The first
 *   label is what a person calls the instance — but only while it stays
 *   unambiguous, hence {@link shortHosts} deciding for the whole set rather
 *   than one host at a time.
 */

import { WinnowError } from './winnow/client';

/** Where an instance stands, the last time this session asked. */
export type HealthState =
  /** Asked, no answer yet. */
  | 'checking'
  /** `/api/capabilities` answered: the sheet on screen is fresh. */
  | 'reachable'
  /** 401/403 — the instance is up and does not know (or allow) this browser. */
  | 'signin'
  /** No answer at all: offline, wrong address, CORS refused, or 5xx. */
  | 'unreachable';

export interface SourceHealth {
  state: HealthState;
  /** What went wrong, in the reader's terms. `null` while fine or checking. */
  reason: string | null;
  /** Round trip of the capabilities call, when it answered. */
  latencyMs: number | null;
  /** When this session last asked. `null` before the first answer. */
  checkedAt: number | null;
}

export const CHECKING: SourceHealth = {
  state: 'checking',
  reason: null,
  latencyMs: null,
  checkedAt: null,
};

/** The health a failed probe leaves behind — the error mapped to a sentence. */
export function healthFromError(err: unknown, host: string, now = Date.now()): SourceHealth {
  const kind = err instanceof WinnowError ? err.kind : null;
  if (kind === 'unauthenticated' || kind === 'forbidden') {
    return {
      state: 'signin',
      reason:
        kind === 'forbidden'
          ? `${host} knows this browser but will not answer for this account.`
          : `${host} does not know this browser. Sign in there, then check again — the session stays in that site's cookie, never here.`,
      latencyMs: null,
      checkedAt: now,
    };
  }
  return {
    state: 'unreachable',
    reason:
      kind === 'unreachable'
        ? `No answer from ${host}. It may be offline, or this browser may not be on its allowed list.`
        : `${host} answered something this app could not read${
            err instanceof Error && err.message ? `: ${err.message}` : '.'
          }`,
    latencyMs: null,
    checkedAt: now,
  };
}

/** The health a successful probe leaves behind. */
export function healthFromAnswer(latencyMs: number, now = Date.now()): SourceHealth {
  return { state: 'reachable', reason: null, latencyMs, checkedAt: now };
}

export interface DocCount {
  projects: number;
  trips: number;
}

export const NO_DOCS: DocCount = { projects: 0, trips: 0 };

/**
 * How many documents each source holds, counted from the two local stores.
 *
 * A document with no `sourceId` was written before the field existed and
 * belongs to this browser — the same rule `groupBySource` applies, kept
 * identical on purpose so the gallery and this screen never disagree.
 */
export function countBySource(
  projects: readonly { sourceId?: string }[],
  trips: readonly { sourceId?: string }[],
  localId = 'local',
): Map<string, DocCount> {
  const counts = new Map<string, DocCount>();
  const bump = (id: string | undefined, key: keyof DocCount) => {
    const at = id ?? localId;
    const current = counts.get(at) ?? { projects: 0, trips: 0 };
    counts.set(at, { ...current, [key]: current[key] + 1 });
  };
  for (const p of projects) bump(p.sourceId, 'projects');
  for (const t of trips) bump(t.sourceId, 'trips');
  return counts;
}

/** "8 projects and 3 trips", "one trip", "nothing yet" — never "0 projects". */
export function describeDocs(count: DocCount): string {
  const parts: string[] = [];
  if (count.projects > 0) {
    parts.push(count.projects === 1 ? 'one project' : `${count.projects} projects`);
  }
  if (count.trips > 0) parts.push(count.trips === 1 ? 'one trip' : `${count.trips} trips`);
  if (parts.length === 0) return 'nothing yet';
  return parts.join(' and ');
}

/**
 * What forgetting a connection actually does, said before it is done.
 *
 * The documents are not deleted: they live on the instance, and connecting
 * again brings them back. What goes is this browser's copy and the ability to
 * save to that host — which is the part a person needs to hear.
 */
export function forgetWarning(host: string, count: DocCount): string {
  const held = describeDocs(count);
  return held === 'nothing yet'
    ? `Forget ${host}? Its media stop showing in the library. Nothing is deleted there.`
    : `Forget ${host}? ${held} came from it — those stay on the instance, and connecting again brings them back, but this browser stops listing and saving them.`;
}

/**
 * The name a host goes by in a tab: its first label, with the port when it
 * carries one (`localhost:5174` is a different instance from `localhost`).
 */
export function shortHost(host: string): string {
  const [name = '', port] = host.split(':');
  const labels = name.split('.').filter(Boolean);
  if (labels.length === 0) return host;
  const first = labels[0] === 'www' && labels.length > 1 ? labels[1] : labels[0];
  // An address is what it is: `192.168.1.4` shortened to `192` names nothing.
  const head = /^\d+$/.test(first) ? name : first;
  return port ? `${head}:${port}` : head;
}

/**
 * Short names for a whole set of hosts, where a collision falls back to the
 * full host for the hosts that collide — `winnow.a.tech` and `winnow.b.tech`
 * both stay long rather than both reading "winnow".
 */
export function shortHosts(hosts: readonly string[]): Map<string, string> {
  const byShort = new Map<string, string[]>();
  for (const host of hosts) {
    const short = shortHost(host);
    byShort.set(short, [...(byShort.get(short) ?? []), host]);
  }
  const out = new Map<string, string>();
  for (const [short, group] of byShort) {
    for (const host of group) out.set(host, group.length === 1 ? short : host);
  }
  return out;
}

File: apple/Packages/AtelierKit/Sources/AtelierKit/Sources/SourceLedger.swift
// What the sources screen needs to say about a source — port of
// `src/shared/sources/source-ledger.ts`. Three jobs, none of which belongs in
// a view:
//
// - **A health state and its sentence.** A connection is not a boolean: an
//   instance can answer, refuse (401 — sign in *there*), or never answer, and
//   each asks for a different verb. The mapping from a `WinnowError` to that
//   state lives here, not in the screen that happens to show it first.
// - **How many documents a source holds**, counted BEFORE forgetting it —
//   "8 projects and 3 trips live there" is the difference between a decision
//   and a surprise.
// - **A short label for a host.** `winnow.steeve.website` truncated in a
//   narrow tab reads as a defect; the first label is what a person calls the
//   instance — but only while it stays unambiguous, hence `shortHosts`
//   deciding for the whole set.

import Foundation

/// Where an instance stands, the last time this session asked.
public enum HealthState: String, Equatable, Sendable, CaseIterable {
    /// Asked, no answer yet.
    case checking
    /// `/api/capabilities` answered: the sheet on screen is fresh.
    case reachable
    /// 401/403 — the instance is up and does not know (or allow) this device.
    case signin
    /// No answer at all: offline, wrong address, a refused origin, or 5xx.
    case unreachable
}

public struct SourceHealth: Equatable, Sendable {
    public var state: HealthState
    /// What went wrong, in the reader's terms. Nil while fine or checking.
    public var reason: String?
    /// Round trip of the capabilities call, when it answered.
    public var latencyMs: Double?
    /// When this session last asked. Nil before the first answer.
    public var checkedAt: Double?

    public init(state: HealthState, reason: String?, latencyMs: Double?, checkedAt: Double?) {
        self.state = state; self.reason = reason; self.latencyMs = latencyMs; self.checkedAt = checkedAt
    }
}

/// The web's `CHECKING`.
public let checkingHealth = SourceHealth(state: .checking, reason: nil, latencyMs: nil, checkedAt: nil)

/// The health a failed probe leaves behind — the error mapped to a sentence.
public func healthFromError(_ error: Error, _ host: String, now: Double) -> SourceHealth {
    let winnow = error as? WinnowError
    if let kind = winnow?.kind, kind == .unauthenticated || kind == .forbidden {
        let reason = kind == .forbidden
            ? "\(host) knows this browser but will not answer for this account."
            : "\(host) does not know this browser. Sign in there, then check again — the session stays in that site's cookie, never here."
        return SourceHealth(state: .signin, reason: reason, latencyMs: nil, checkedAt: now)
    }
    if winnow?.kind == .unreachable {
        return SourceHealth(state: .unreachable,
                            reason: "No answer from \(host). It may be offline, or this browser may not be on its allowed list.",
                            latencyMs: nil, checkedAt: now)
    }
    let message = winnow?.message ?? (error as? LocalizedError)?.errorDescription
    let tail = message.map { $0.isEmpty ? "." : ": \($0)" } ?? "."
    return SourceHealth(state: .unreachable, reason: "\(host) answered something this app could not read\(tail)",
                        latencyMs: nil, checkedAt: now)
}

/// The health a successful probe leaves behind.
public func healthFromAnswer(_ latencyMs: Double, now: Double) -> SourceHealth {
    SourceHealth(state: .reachable, reason: nil, latencyMs: latencyMs, checkedAt: now)
}

public struct DocCount: Equatable, Sendable {
    public var projects: Int
    public var trips: Int
    public var rolls: Int

    public init(projects: Int = 0, trips: Int = 0, rolls: Int = 0) {
        self.projects = projects; self.trips = trips; self.rolls = rolls
    }
}

/// The web's `NO_DOCS`.
public let noDocs = DocCount()

/// How many documents each source holds, counted from the local stores — one
/// `sourceId` per document. A document with none was written before the field
/// existed and belongs to this device: the rule `groupBySource` applies, kept
/// identical so the gallery and this screen never disagree.
public func countBySource(projects: [String?] = [], trips: [String?] = [], rolls: [String?] = [],
                          localId: String = defaultSourceId) -> [String: DocCount] {
    var counts: [String: DocCount] = [:]
    for id in projects { counts[id ?? localId, default: noDocs].projects += 1 }
    for id in trips { counts[id ?? localId, default: noDocs].trips += 1 }
    for id in rolls { counts[id ?? localId, default: noDocs].rolls += 1 }
    return counts
}

/// "8 projects, 3 trips and 2 rolls", "one trip", "nothing yet" — never "0 projects".
public func describeDocs(_ count: DocCount) -> String {
    var parts: [String] = []
    func say(_ n: Int, _ noun: String) {
        if n > 0 { parts.append(n == 1 ? "one \(noun)" : "\(n) \(noun)s") }
    }
    say(count.projects, "project")
    say(count.trips, "trip")
    say(count.rolls, "roll")
    if parts.isEmpty { return "nothing yet" }
    if parts.count == 1 { return parts[0] }
    return "\(parts.dropLast().joined(separator: ", ")) and \(parts[parts.count - 1])"
}

/// What forgetting a connection actually does, said before it is done. The
/// documents are not deleted — they live on the instance, and connecting again
/// brings them back; what goes is this device's copy and the ability to save.
public func forgetWarning(_ host: String, _ count: DocCount) -> String {
    let held = describeDocs(count)
    return held == "nothing yet"
        ? "Forget \(host)? Its media stop showing in the library. Nothing is deleted there."
        : "Forget \(host)? \(held) came from it — those stay on the instance, and connecting again brings them back, but this browser stops listing and saving them."
}

/// The name a host goes by in a tab: its first label (past a `www`), with the
/// port when it carries one (`localhost:5174` is another instance than
/// `localhost`); an IP address stays whole.
public func shortHost(_ host: String) -> String {
    let bits = host.components(separatedBy: ":")
    let name = bits.first ?? ""
    let port = bits.count > 1 ? bits[1] : ""
    let labels = name.split(separator: ".").map(String.init).filter { !$0.isEmpty }
    guard !labels.isEmpty else { return host }
    let first = labels[0] == "www" && labels.count > 1 ? labels[1] : labels[0]
    let allDigits = !first.isEmpty && first.utf8.allSatisfy { $0 >= 48 && $0 <= 57 }
    let head = allDigits ? name : first
    return port.isEmpty ? head : "\(head):\(port)"
}

/// Short names for a whole set of hosts, where a collision falls back to the
/// full host for the hosts that collide — `winnow.a.tech` and `winnow.b.tech`
/// both stay long rather than both reading "winnow".
public func shortHosts(_ hosts: [String]) -> [String: String] {
    var byShort: [String: [String]] = [:]
    for host in hosts { byShort[shortHost(host), default: []].append(host) }
    var out: [String: String] = [:]
    for (short, group) in byShort {
        for host in group { out[host] = group.count == 1 ? short : host }
    }
    return out
}
